proposals/0030-markdoc.md

```diff
@@ -0,0 +1,295 @@
+- **Benchmark Markdoc performance against Markdown and MDX** at 1k and 10k documents. This addresses problem (1) from the previous section.
+
+# Non-goals
+
+- Non-goal: **ESM import and `src/pages/` support** for Markdoc files. [See discussion](https://github.com/withastro/roadmap/discussions/478#discussioncomment-4975072) for context.
+- Out-of-scope: **Allowing the `.md` extension.** This would mean overriding Astro's `.md` renderer, which is tightly coupled to remark and your `markdown` configuration options today. We agree using `.md` for Markdoc is a common use case, and deserves a separate proposal to make Astro's Markdown rendering flexible.
+- Future: **A solution for client-rendered UI components.** Unlike MDX, Markdoc doesn't have a concept of directives, and our compiler doesn't have a clear way to dynamically render client-side components ([see challenges](https://gist.github.com/bholmesdev/491fa440efa8bd73d410d66c0c2143c2)). We will recommend users wrap their components in an Astro component to apply the `client:` directive.
+- Future: **Full alignment with Markdown and MDX** rendered result. Namely, the computed `headings` property (which can be tackled in future releases) and frontmatter manipulation via remark (since remark is incompatible with Markdoc).
+
+# Detailed Design
+
+## API
+
+Markdoc support will be specific to content collections. This means Markdoc files can be included alongside other content collection entries via the `.mdoc` extension. Entries can also be queried using the same `getCollection()` and `getEntryBySlug()` APIs for Markdown and MDX.
+
+### Rendering Markdoc content
+
+Users can render contents and configure Markdoc via the `<Content />` component. This will be exposed from the `render()` result, and feature two props:
+
+- `config?: import('@markdoc/markdoc').Config`: An (optional) Markdoc config to be used [during the transformation step](https://markdoc.dev/docs/render#transform). This includes configuration for Markdoc tags and variables.
+- `components?: Record<string, ComponentRenderer>`: An (optional) mapping from Markdoc tags or elements to Astro components.
+
+We expect users to share a common Markdoc `config` and `components` setup throughout their application. Because of this, we will recommend creating a utility component to encapsulate that config.
+
+For example, say a user wants to share config specific to their `blog` collection. This `BlogContent` component will add support for an `{% aside %}` tag across blog entries:
+
+```astro
+---
+// src/components/BlogContent.astro
+import Aside from './Aside.astro';
+import type { CollectionEntry } from 'astro:content';
+
+type Props = {
+	entry: CollectionEntry<'blog'>;
+};
+
+const { entry } = Astro.props;
+const { Content } = await entry.render();
+---
+
+<Content
+	config={{
+		tags: {
+			aside: {
+				render: 'Aside',
```

Comment:
this use of a string is the only issue I have with in the current API. Keeping this string value matching the JS variable name is always an annoying and not-super-obvious API. 

Is there any way to pass `Aside` as a reference here, and then skip the `components={}` step entirely? Something like:

```
<Content
	config={{
		tags: {
			aside: {
				component: Aside,
				attributes: {
				    type: { type: String },
					title: { type: String },
				},
			},
		},
	}}
/>
```

I can't overstate how poor a DX it is to maintain a string value referencing a variable name in the same file like this. TS type checking may help here but removing the problem entirely would be better.

Reply:
Probably related to the conversation Ben & I had here: https://github.com/withastro/roadmap/issues/496#issuecomment-1446402335

Reply:
I understand the DX concern here! However, I'm opposed to changing the shape of `config` in any way before launch. This follows Markdoc's transform config 1-1, and it feels wrong to modify certain aspects of that config for our purposes. That would feel like modifying the shape of the Vite config when using Astro. 

I also understand your concern over using plain strings, though this was an intentional choice by the Markdoc team to allow rendering a plain HTML element instead of a component (i.e. `'aside'`). I agree we could explore modifications to Markdoc's config in the future, though I'd much prefer to follow the precedent set by Markdoc's React renderer:
- `config` used for the transformation step, with no reference to rendered components
- `components` used for the render step, now with reference to rendered components

Reply:
Update: we may need to move the `config` prop off of the `Content` component and towards some global export as well to address @delucis' "run transforms at build time" suggestion. This could clarify the separation that's fairly opaque in this example.

Reply:
Discussed offline! This API has been redesigned to colocate config in your `astro.config.mjs`. This should make the string -> component map tradeoff more usable for an MVP. We plan to improve from here by experimenting with the [NextJS integration's approach](https://markdoc.dev/docs/nextjs) or a reserved folder to autowire this config (i.e. `embeds/`)